src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,138 @@
+
+import torch
+
+from torchcodec.decoders import FrameBatch, SimpleVideoDecoder
+
+
+def _validate_params(
+    *, decoder, num_clips, num_frames_per_clip, num_indices_between_frames
+):
+    if len(decoder) < 1:
+        raise ValueError(
+            f"Decoder must have at least one frame, found {len(decoder)} frames."
+        )
+
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+
+def _validate_sampling_range(
+    *, sampling_range_start, sampling_range_end, num_frames, clip_span
+):
+    if sampling_range_start < 0:
+        sampling_range_start = num_frames + sampling_range_start
+
+    # TODO: or max(sampling_range_start, num_frames - 1)?
+    sampling_range_start = sampling_range_start % num_frames
+
+    if sampling_range_end is None:
+        sampling_range_end = num_frames - clip_span + 1
+        if sampling_range_start >= sampling_range_end:
+            raise ValueError(
+                f"We determined that sampling_range_end should be {sampling_range_end}, "
+                "but it is smaller than or equal to sampling_range_start "
+                f"({sampling_range_start})."
+            )
+    else:
+        if sampling_range_end < 0:
+            # Support negative values so that -1 means last frame.
+            sampling_range_end = num_frames + sampling_range_end
+        if sampling_range_start >= sampling_range_end:
+            raise ValueError(
+                f"sampling_range_start ({sampling_range_start}) must be smaller than "
+                f"sampling_range_end ({sampling_range_end})."
+            )
+
+    return sampling_range_start, sampling_range_end
+
+
+def _get_clip_span(*, num_indices_between_frames, num_frames_per_clip):
+    """Return the span of a clip, i.e. the number of frames (or indices)
+    between the first and last frame in the clip, both included.
+
+    This isn't the same as the number of frames in a clip!
+    Example: f means a frame in the clip, x means a frame excluded from the clip
+    num_frames_per_clip = 4
+    num_indices_between_frames = 1, clip = ffff      , span = 4
+    num_indices_between_frames = 2, clip = fxfxfxf   , span = 7
+    num_indices_between_frames = 3, clip = fxxfxxfxxf, span = 10
+    """
+    return num_indices_between_frames * (num_frames_per_clip - 1) + 1
+
+
+def clips_at_random_indices(
+    decoder: SimpleVideoDecoder,
+    *,
+    num_clips: int = 1,
+    num_frames_per_clip: int = 1,
+    num_indices_between_frames: int = 1,
+    sampling_range_start: int = 0,
+    sampling_range_end: Optional[int] = None,  # interval is [start, end).
+) -> List[FrameBatch]:
```

Comment:
Did we agree to return a List here? I thought some users had a need to get a single tensor back? Would those users just stack it manually? Note that stacking could be an expensive operation.

Reply:
When we discussed this, we decided to "let the implementation decide".
I do agree that stacked tensors make more sense in general, but we also don't want to stack on behalf of the user as it may introduce an unnecessary copy.
The current implementation leads to a list, but it's possible by implementing the "better frame shuffling" strategy (left as a TODO at the bottom of this function), we'll end up with a stack. This is still TBD.

Reply:
Decoding frames to a single tensor doesn't involves a copy -- and that can be done for max speed if we want.

Stacking them after the fact is inefficient.